src/Utilities/Functional.hpp

```diff
@@ -0,0 +1,255 @@
+ * \brief creates binary functors from functions
+ *
+ * \details
+ * creates a struct with name STRUCTNAME which calls function FUNC for any
+ * parameter. It can be supplied an explicit type,
+ * e.g. `STRUCTNAME<double,double>{}(val1,val2)`, or can infer the type if no
+ * type is supplied, e.g. `STRUCTNAME<>{}(val1,val2)`
+ *
+ */
+#define MAKE_BINARY_STRUCT(STRUCTNAME, FUNC)          \
+  template <typename T1 = void, typename T2 = void>   \
+  struct STRUCTNAME {                                 \
+    auto operator()(const T1& arg1, const T2& arg2) { \
+      return (FUNC(arg1, arg2));                      \
+    }                                                 \
+  };                                                  \
+  template <>                                         \
+  struct STRUCTNAME<void, void> {                     \
+    template <typename T1, typename T2>               \
+    auto operator()(const T1& arg1, const T2& arg2) { \
+      return (FUNC(arg1, arg2));                      \
+    }                                                 \
+  };
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief creates binary assign functors from functions
+ *
+ * \details
+ * creates a struct with name STRUCTNAME which calls function FUNC for any
+ * parameter. It can be supplied an explicit type,
+ * e.g. `STRUCTNAME<double>{}(val)`, or can infer the type if no type is
+ * supplied, e.g. `STRUCTNAME<>{}(val)`
+ *
+ */
+#define MAKE_BINARY_INPLACE_STRUCT(STRUCTNAME, FUNC) \
+  template <typename T1 = void, typename T2 = void>  \
+  struct STRUCTNAME {                                \
+    auto operator()(T1& arg1, const T2& arg2) {      \
+      return (arg1 = FUNC(arg1, arg2));              \
+    }                                                \
+  };                                                 \
+  template <>                                        \
+  struct STRUCTNAME<void, void> {                    \
+    template <typename T1, typename T2>              \
+    auto operator()(T1& arg1, const T2& arg2) {      \
+      return (arg1 = FUNC(arg1, arg2));              \
+    }                                                \
+  };
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief defines the step function for arithmetic types
+ */
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
+constexpr inline T step_function(const T& arg) {
+  return (arg > 0) ? 1 : 0;
+}
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief defines the inverse square-root (\f$1/\sqrt{x}\f$) for arithmetic
+ * types and complex
+ */
+template <typename T,
+          Requires<std::is_arithmetic<T>::value ||
+                   cpp17::is_same_v<T, std::complex<double> > > = nullptr>
+constexpr inline T invsqrt(const T& arg) {
+  return 1.0 / sqrt(arg);
+}
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief defines the inverse cube-root (\f$1/{}^3\sqrt{x}\f$) for arithmetic
+ * types
+ */
+template <typename T, Requires<std::is_arithmetic<T>::value> = nullptr>
```

Comment:
I want to avoid defining the template for any types apart from arithmetic types for safety (e.g. regardless of how screwed up an include set is, this shouldn't run on a vector), because other definitions with the same name exist. Is there a better way to enforce that?

Reply:
The compilation will fail because the wrapped function doesn't exist for other types, so the desired behavior is already present :)

Reply:
true for step_function, not true for invsqrt. The sqrt function is defined for other types, so the interior of the function should work fine for DataVectors and (might?) shadow the blaze version, which is doubtlessly better optimized than just evaluating 1.0/sqrt(x).